File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/voice/voice_component_view_model_base.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:runanywhere/runanywhere.dart' as sdk;

import 'package:runanywhere_ai/features/models/model_list_view_model.dart';
import 'package:runanywhere_ai/features/models/model_types.dart';

/// Shared lifecycle base for the single-component Voice ViewModels
/// (STT / TTS / VAD). Mirrors iOS `VoiceComponentViewModelBase.swift`:
/// it owns the idempotent initialize/subscribe/cleanup plumbing every
/// voice-component screen needs so each concrete ViewModel only declares
/// its component identity and how it maps an SDK event onto its state.
///
/// `VoiceAgentViewModel` is intentionally NOT built on this base: it
/// orchestrates the full STT -> LLM -> TTS pipeline and consumes
/// per-component event streams plus lifecycle snapshots, a richer pattern
/// than the single-component load/unload tracking captured here.
abstract class VoiceComponentViewModelBase extends ChangeNotifier {
  VoiceComponentViewModelBase();

  // --- Published model state -----------------------------------------------

  /// Framework of the currently loaded model (for the status banner).
  LLMFramework? selectedFramework;

  /// Display name of the currently loaded model.
  String? selectedModelName;

  /// Registry id of the currently loaded model.
  String? selectedModelId;

  /// Last error to surface in the UI, or null.
  String? errorMessage;

  // --- Subscription / idempotency state -------------------------------------

  StreamSubscription<sdk.ModelLifecycleChange>? _lifecycleSubscription;
  bool _isInitialized = false;
  bool _hasSubscribedToSDKEvents = false;
  bool _disposed = false;

  // --- Component identity (override in subclasses) --------------------------

  /// SDK component this ViewModel tracks (e.g. `SDK_COMPONENT_STT`).
  sdk.SDKComponent get component;

  /// Event category that carries this component's lifecycle events.
  sdk.EventCategory get eventCategory;

  /// Model category used to query the initial loaded model.
  ModelCategory get modelCategory;

  /// Whether a model is currently selected/loaded for this component.
  bool get hasModelSelected => selectedModelId != null;

  /// Surface an error produced outside the ViewModel (e.g. a denied
  /// permission request, which the view owns because it needs BuildContext).
  void reportError(String message) {
    errorMessage = message;
    notify();
  }

  // --- Initialization --------------------------------------------------------

  /// Claim the one-time initialization guard. Subclasses call this at the
  /// top of their `initialize()`; when it returns `true` they own the
  /// remaining setup (subscriptions, initial-state query).
  @protected
  bool beginInitialization() {
    if (_isInitialized) {
      debugPrint('Voice component view model already initialized, skipping');
      return false;
    }
    _isInitialized = true;
    return true;
  }

  // --- Model loading ---------------------------------------------------------

  /// Modality-specific canonical SDK load call
  /// (`stt.load` / `tts.loadVoice` / `vad.loadModel`).
  @protected
  Future<void> performLoad(ModelInfo model);

  /// Load a model for this component via the canonical SDK entry point.
  /// Shared by every voice-component screen; subclasses wrap it only to
  /// toggle their busy flag.
  Future<bool> loadModel(ModelInfo model) async {
    debugPrint('Loading $component model: ${model.name}');
    errorMessage = null;
    notify();

    try {
      await performLoad(model);
    } catch (e) {
      debugPrint('Model load failed: $e');
      errorMessage = 'Failed to load model: $e';
      notify();
      return false;
    }
    applyLoadedModel(model);
    debugPrint('Model loaded: ${model.id}');
    return true;
  }

  // --- SDK event subscription ------------------------------------------------

  /// Subscribe to the SDK's unified model-lifecycle stream, filtered to this
  /// component. Idempotent. Mirrors iOS `subscribeToSDKEvents()`.
  @protected
  void subscribeToSDKEvents() {
    if (_hasSubscribedToSDKEvents) {
      debugPrint('Already subscribed to SDK events, skipping');
      return;
    }
    _hasSubscribedToSDKEvents = true;

    _lifecycleSubscription = sdk.RunAnywhere.events.modelLifecycle.listen((
      change,
    ) {
      if (change.component != component &&
          change.event.category != eventCategory) {
        return;
      }
      switch (change.kind) {
        case sdk.ModelLifecycleChangeKind.loaded:
          // Resolve from the event payload rather than doing extra snapshot
          // work inside the lifecycle handler. The change carries the model
          // id; resolve the rest from the catalog and apply through
          // `applyLoadedModel` so the subclass
          // overrides (STT live-mode, TTS system-voice) still run.
          applyLoadedModel(resolveLoadedModel(change.modelId));
        case sdk.ModelLifecycleChangeKind.unloaded:
          clearLoadedModel();
          debugPrint('Voice component model unloaded');
      }
    });
  }

  // --- Initial model state ----------------------------------------------------

  /// Apply the SDK's current-model snapshot at cold start.
  @protected
  Future<void> checkInitialModelState() =>
      applyCurrentModelSnapshot('already loaded');

  /// Resolve the current model for this modality via the SDK snapshot and
  /// apply it to published state. Used ONLY at cold start
  /// ([checkInitialModelState]); a one-shot `current()` outside an event
  /// handler is safe. The lifecycle-loaded listener resolves from the event
  /// payload instead so it does not perform redundant SDK snapshot work.
  @protected
  Future<void> applyCurrentModelSnapshot(String reason) async {
    final result = await sdk.RunAnywhere.modelLifecycle.current(
      sdk.CurrentModelRequest(
        category: modelCategory,
        includeModelMetadata: true,
      ),
    );
    if (_disposed || !result.found) return;

    // When the snapshot omits model metadata, `model.id` is empty; fall back
    // to the top-level `modelId` so the id is always populated.
    final model = result.model.deepCopy();
    if (model.id.isEmpty) {
      model.id = result.modelId;
    }
    applyLoadedModel(model);
    debugPrint('Voice component model $reason: ${model.id}');
  }

  /// Apply a model resolved at startup (or via an event) to published state.
  /// Default uses the model's own name/framework; override when a modality
  /// needs different resolution (e.g. catalog lookup or voice-id display).
  @protected
  void applyLoadedModel(ModelInfo model) {
    selectedModelId = model.id;
    selectedModelName = model.name.isNotEmpty ? model.name : model.id;
    selectedFramework = model.framework;
    notify();
  }

  /// Resolve display name/framework through the shared model catalog when
  /// available (mirrors the iOS STT/VAD `applyLoadedModel` overrides).
  @protected
  void applyLoadedModelFromCatalog(ModelInfo model) {
    selectedModelId = model.id;
    final matches = ModelListViewModel.shared.availableModels.where(
      (m) => m.id == model.id,
    );
    if (matches.isNotEmpty) {
      final match = matches.first;
      selectedModelName = match.name;
      selectedFramework = match.framework;
    } else {
      selectedModelName = model.name.isNotEmpty ? model.name : model.id;
      selectedFramework = model.framework;
    }
    notify();
  }

  /// Resolve the full [ModelInfo] for a just-loaded [modelId] from the shared
  /// catalog, falling back to an id-only model when the catalog has no entry.
  /// Used by the lifecycle-loaded listener, which only carries the id — this
  /// avoids re-querying the SDK (which would loop) while still giving the
  /// subclass `applyLoadedModel` overrides a populated name/framework.
  @protected
  ModelInfo resolveLoadedModel(String modelId) {
    final matches = ModelListViewModel.shared.availableModels.where(
      (m) => m.id == modelId,
    );
    if (matches.isNotEmpty) return matches.first;
    return ModelInfo()..id = modelId;
  }

  /// Clear published state for an unloaded model.
  @protected
  void clearLoadedModel() {
    selectedModelId = null;
    selectedModelName = null;
    selectedFramework = null;
    notify();
  }

  // --- Cleanup ----------------------------------------------------------------

  /// Notify listeners unless the ViewModel was already disposed (async SDK
  /// callbacks can land after the screen is gone).
  @protected
  void notify() {
    if (_disposed) return;
    notifyListeners();
  }

  /// Tear down subscriptions and reset the idempotency guards so the
  /// ViewModel can be re-initialized.
  @protected
  void cleanupBase() {
    unawaited(_lifecycleSubscription?.cancel());
    _lifecycleSubscription = null;
    _isInitialized = false;
    _hasSubscribedToSDKEvents = false;
  }

  @override
  void dispose() {
    _disposed = true;
    cleanupBase();
    super.dispose();
  }
}

File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Voice/VoiceComponentViewModelBase.swift
//
//  VoiceComponentViewModelBase.swift
//  RunAnywhereAI
//
//  Shared lifecycle base for the single-component Voice ViewModels
//  (STT / TTS / VAD). It owns the idempotent initialize/subscribe/cleanup
//  plumbing that every voice-component screen needs so each concrete
//  ViewModel only declares its component identity and how it maps an
//  SDK event onto its published model state.
//
//  VoiceAgentViewModel is intentionally NOT built on this base: it
//  orchestrates the full STT→LLM→TTS pipeline and consumes per-component
//  event streams plus lifecycle snapshots, a richer pattern than the
//  single-component load/unload tracking captured here.
//

import Foundation
import RunAnywhere
import Combine
import os

/// Base class for ViewModels that observe a single SDK voice component's
/// load/unload lifecycle (STT, TTS, VAD).
///
/// Subclasses must override `component`, `eventCategory`, and `modelCategory`
/// to identify their modality. The base provides the shared
/// `loadModel(from:)` entry point, the idempotent `subscribeToSDKEvents()`
/// (built on the SDK's typed `modelLifecycle` stream), and
/// `checkInitialModelState()` / `cleanupBase()` helpers shared by all three.
@MainActor
class VoiceComponentViewModelBase: ObservableObject {
    // MARK: - Published Model State

    @Published var selectedFramework: InferenceFramework?
    @Published var selectedModelName: String?
    @Published var selectedModelId: String?
    @Published var errorMessage: String?

    // MARK: - Subscription / Idempotency State

    let logger: Logger
    var cancellables = Set<AnyCancellable>()
    private var isInitialized = false
    private var hasSubscribedToSDKEvents = false

    // MARK: - Component Identity (override in subclasses)

    /// SDK component this ViewModel tracks (e.g. `.stt`).
    var component: RASDKComponent { .unspecified }

    /// Event category that carries this component's lifecycle events.
    var eventCategory: RAEventCategory { .unspecified }

    /// Model category used to query the initial loaded model.
    var modelCategory: RAModelCategory { .unspecified }

    // MARK: - Initialization

    init(loggerCategory: String) {
        self.logger = Logger(subsystem: "com.runanywhere", category: loggerCategory)
    }

    /// Claim the one-time initialization guard. Subclasses call this at the
    /// top of their `initialize()`; when it returns `true` they own the
    /// remaining setup (permissions, subscriptions, initial-state query) and
    /// may order it however the modality requires.
    ///
    /// - Returns: `true` when this is the first initialization, `false` when
    ///   already initialized and the caller should bail out.
    func beginInitialization() -> Bool {
        guard !isInitialized else {
            logger.debug("Voice component view model already initialized, skipping")
            return false
        }
        isInitialized = true
        return true
    }

    // MARK: - Model Loading

    /// Load a model for this component via the canonical SDK entry point.
    /// Shared by every voice-component screen; subclasses wrap it only to
    /// toggle their busy flag.
    @discardableResult
    func loadModel(from model: RAModelInfo) async -> Bool {
        logger.info("Loading \(String(describing: self.component)) model: \(model.name)")
        errorMessage = nil

        var request = RAModelLoadRequest()
        request.modelID = model.id
        request.category = modelCategory
        let result = await RunAnywhere.loadModel(request)
        guard result.success else {
            logger.error("Model load failed: \(result.errorMessage)")
            errorMessage = "Failed to load model: \(result.errorMessage)"
            return false
        }
        applyLoadedModel(model)
        logger.info("Model loaded: \(model.id)")
        return true
    }

    // MARK: - SDK Event Subscription

    func subscribeToSDKEvents() {
        guard !hasSubscribedToSDKEvents else {
            logger.debug("Already subscribed to SDK events, skipping")
            return
        }
        hasSubscribedToSDKEvents = true

        // The SDK's typed lifecycle stream folds the three native load/unload
        // channels into one publisher; the base just applies it to this
        // component's published state.
        RunAnywhere.events.modelLifecycle
            .receive(on: DispatchQueue.main)
            .sink { [weak self] change in
                // Defer state modifications to avoid "Publishing changes
                // within view updates" warnings.
                Task { @MainActor in
                    guard let self else { return }
                    guard change.component == self.component
                        || change.event.category == self.eventCategory else { return }
                    switch change.kind {
                    case .loaded:
                        guard !change.modelID.isEmpty else {
                            self.logger.debug("Ignoring loaded lifecycle event without model id")
                            return
                        }
                        let previousModelId = self.selectedModelId
                        let model = self.resolveLoadedModel(change.modelID)
                        self.applyLoadedModel(model)
                        if previousModelId != model.id {
                            self.logger.info("Voice component model loaded: \(model.id)")
                        }
                    case .unloaded:
                        self.clearLoadedModel()
                        self.logger.info("Voice component model unloaded")
                    }
                }
            }
            .store(in: &cancellables)
    }

    // MARK: - Initial Model State

    func checkInitialModelState() async {
        applyCurrentModelSnapshot(reason: "already loaded")
    }

    /// Resolve the current model for this modality via the SDK snapshot and
    /// apply it to published state. Shared by `checkInitialModelState()` (cold
    /// start). Lifecycle-loaded events resolve from the event's model id to
    /// avoid a current-model query feeding back into the lifecycle stream.
    func applyCurrentModelSnapshot(reason: String) {
        var request = RACurrentModelRequest()
        request.category = modelCategory
        let snapshot = RunAnywhere.currentModel(request)
        guard snapshot.found else { return }

        // When the snapshot omits model metadata, `model.id` is empty; fall
        // back to the top-level `modelID` so the id is always populated.
        var model = snapshot.model
        if model.id.isEmpty {
            model.id = snapshot.modelID
        }
        applyLoadedModel(model)
        logger.info("Voice component model \(reason): \(model.id)")
    }

    /// Resolve the full model metadata for a lifecycle event without calling
    /// back into the SDK current-model API from inside the event handler.
    func resolveLoadedModel(_ modelId: String) -> RAModelInfo {
        if let matchingModel = ModelListViewModel.shared.availableModels.first(where: { $0.id == modelId }) {
            return matchingModel
        }
        var model = RAModelInfo()
        model.id = modelId
        return model
    }

    /// Apply a model resolved at startup (or via an event) to published state.
    /// Default resolves the display name from the id; override when a modality
    /// needs different name/framework resolution (e.g. catalog lookup).
    func applyLoadedModel(_ model: RAModelInfo) {
        selectedModelId = model.id
        selectedModelName = model.name.modelNameFromID()
        selectedFramework = model.framework
    }

    /// Clear published state for an unloaded model.
    func clearLoadedModel() {
        selectedModelId = nil
        selectedModelName = nil
        selectedFramework = nil
    }

    // MARK: - Cleanup

    /// Tear down subscriptions and reset the idempotency guards so the
    /// ViewModel can be re-initialized. Subclasses call this from their
    /// `cleanup()` after releasing any modality-specific resources.
    func cleanupBase() {
        cancellables.removeAll()
        isInitialized = false
        hasSubscribedToSDKEvents = false
    }
}
